File: b-5639.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<int> tree;

void find(int s, int e)
{
    if(s >= e)
    {
        if(s == e) printf("%d\n", tree[s]);
        return;
    }
    int idx = upper_bound(tree.begin() + s, tree.begin() + e + 1, tree[s]) - tree.begin();
    find(s + 1, idx - 1);
    find(idx, e);
    printf("%d\n", tree[s]);
}
int main()
{
    int tmp;
    while(1){
        cin >> tmp;
        if(cin.eof()) break;
        tree.push_back(tmp);
    }

    find(0, tree.size() - 1);
    return 0;
}

File: b-5639.py
import sys
from bisect import bisect_right
lines = sys.stdin.readlines()
tree = [int(n) for n in lines]

def find(s, e):
    if(s >= e):
        if(s == e):
            print(tree[s])
        return
    idx = bisect_right(tree, tree[s])
    find(s + 1, idx - 1)
    find(idx, e)
    print(tree[s])

print(tree)
find(0, len(tree) - 1)